fhir-smart/src/main/java/com/ibm/fhir/smart/AuthzPolicyEnforcementPersistenceInterceptor.java

```diff
@@ -0,0 +1,268 @@
+            String refVal = reference.getReference().getValue();
+            if (refVal != null && refVal.startsWith(PATIENT_REF_PREFIX)) {
+                return refVal.substring(PATIENT_REF_PREFIX.length());
+            }
+        } else if (log.isLoggable(Level.FINE)){
+            log.fine("Skipping non-patient / non-relative reference: '" + reference + "'");
+        }
+
+        return null;
+    }
+
+    /**
+     * @param permission
+     * @param requiredPermission
+     * @return true if the grantedPermission includes the requiredPermission; otherwise false
+     */
+    private boolean hasPermission(Permission grantedPermission, Permission requiredPermission) {
+        if (grantedPermission == Permission.ALL) {
+            return true;
+        } else {
+            return grantedPermission == requiredPermission;
+        }
+    }
+
+    private String getAccessToken() throws FHIRPersistenceInterceptorException {
+        List<String> list = FHIRRequestContext.get().getHttpHeaders().get("Authorization");
+        if (list.size() != 1) {
+            throw new FHIRPersistenceInterceptorException("Request must contain exactly one Authorization header.");
+        }
+        String header = list.get(0);
+
+        if (!header.startsWith(BEARER_TOKEN_PREFIX)) {
+            throw new FHIRPersistenceInterceptorException("Authorization header must carry a Bearer token");
+        }
+
+        return header.substring(BEARER_TOKEN_PREFIX.length()).trim();
+    }
+
+    private List<Scope> getScopesFromToken(DecodedJWT jwt) {
+        Claim claim = jwt.getClaim("scope");
+        if (claim.isNull()) {
+            throw new IllegalArgumentException("Authorization token is missing 'scope' claim");
+        }
+
+        List<String> scopeStrings;
+        String claimString = claim.asString();
+
+        if (claimString != null) {
+            scopeStrings = Arrays.asList(claim.asString().split("\\s+"));
+        } else {
+            log.fine("Found scope claim was expected to be a string but is not; processing as a list");
+            scopeStrings = claim.asList(String.class);
+        }
+
+        return scopeStrings.stream()
+                .filter(s -> s.matches(Scope.SCOPE_STRING_REGEX))
+                .map(s -> new Scope(s))
+                .collect(Collectors.toList());
+    }
+
+    private List<String> getPatientIdFromToken(DecodedJWT jwt) {
+        Claim claim = jwt.getClaim("patient_id");
+        if (claim.isNull()) {
+            throw new IllegalArgumentException("Authorization token is missing 'patient_id' claim");
+        }
+
+        String patientId = claim.asString();
+        if (patientId == null) {
+            log.fine("Found patient_id claim was expected to be a string but is not; processing as a list");
+            return claim.asList(String.class);
+        }
+
+        return Collections.singletonList(patientId);
+    }
```

Comment:
Are you anticipating more than one id?

Reply:
Wanted to leave the door open for it, but could refactor to assume a single patientId is in-scope if desired.

Reply:
No refactoring. This was a question given the multiple potential ids.